### PR TITLE
🐯 Integrate Liger GRPO Loss to GRPO Trainer

### PR DESCRIPTION
**Pull request by shivam15s**
# What does this PR do?
Integrates LigerFusedLinearGRPOLoss to GRPOTrainer

> Currently works with DP and DDP. Will work on FSDP support in a subsequent PR. 

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [ ] Did you read the [contributor guideline](https://github.com/huggingface/trl/blob/main/CONTRIBUTING.md#create-a-pull-request),
      Pull Request section?
- [ ] Was this discussed/approved via a GitHub issue? Please add a link
      to it if that's the case.
- [ ] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/trl/tree/main/docs).
- [x] Did you write any new necessary tests?

## Who can review?

Anyone in the community is free to review the PR once the tests have passed. Feel free to tag
members/contributors who may be interested in your PR.

### CODE REVIEW

File: trl/trainer/grpo_config.py [hunk 1]
```diff
@@ -378,6 +380,10 @@ class GRPOConfig(TrainingArguments):
             "vLLM, you should now use the `enable_prefix_caching` parameter in the vLLM server configuration."
         },
     )
+    use_liger_loss: bool = field(
+        default=False,
+        metadata={"help": "Whether to use the Liger GRPO loss."},
+    )
```
Comment (qgallouedec) [inline trl/trainer/grpo_config.py:386]: Could you move this into the "Parameters that control the training"

File: trl/trainer/grpo_trainer.py [hunk 2]
```diff
@@ -940,17 +969,53 @@ def _generate_and_score_completions(
...
+        return loss
+
     @profiling_decorator
     def compute_loss(self, model, inputs, return_outputs=False, num_items_in_batch=None):
         if return_outputs:
             raise ValueError("The GRPOTrainer does not support returning outputs")
-        # Compute the per-token log probabilities for the model
+        if self.use_liger_loss:
+            # Compute the loss using the liger grpo loss
+            return self.compute_liger_loss(model, inputs)
```
Comment (qgallouedec) [inline trl/trainer/grpo_trainer.py:1016]:
```suggestion
        if self.use_liger_loss:
            # Compute the loss using the liger grpo loss
            return self._compute_loss_with_liger(model, inputs)
        else:
            return self._compute_loss(model, inputs)
```

what do you think?

File: trl/trainer/grpo_trainer.py [hunk 3]
```diff
@@ -602,13 +621,23 @@ def _enable_gradient_checkpointing(self, model: PreTrainedModel, args: GRPOConfi
 
         return model
 
+    @profiling_decorator
+    def _get_hidden_states(self, model, input_ids, attention_mask, logits_to_keep=None):
+        # need to unwrap the model to get the hidden states
+        # intended to work with DP, DDP, FSDP
+        unwrapped_model = self.accelerator.unwrap_model(model)
+        hidden_states = unwrapped_model.model(input_ids=input_ids, attention_mask=attention_mask)[0]
```
Comment (qgallouedec) [inline trl/trainer/grpo_trainer.py:629]: here you get the logits, not the hidden_states, don't you?

You want to use this instead:

```suggestion
        hidden_states = unwrapped_model.model(input_ids=input_ids, attention_mask=attention_mask, output_hidden_states=True).hidden_states
```

Comment (shivam15s) [reply trl/trainer/grpo_trainer.py:629]: just need the last hidden state of the model

Comment (shivam15s) [reply trl/trainer/grpo_trainer.py:629]: as we call the unwrapped_model.model (it is calling the base model which returns BaseModelOutput)

Comment (shivam15s) [reply trl/trainer/grpo_trainer.py]: refactored to use last_hidden_state option explicitly

**Comment by shivam15s** [comment]
liger-kernel [v0.5.6](https://github.com/linkedin/Liger-Kernel/tree/v0.5.6) is out which has the changes needed for grpo, so we can officially test this integration!

File: tests/slow/test_grpo_slow.py [hunk 4]
```diff
@@ -0,0 +1,82 @@
...
+    @require_liger_kernel
+    def test_training_with_liger_grpo_loss(self, model_name):
+        with tempfile.TemporaryDirectory() as tmp_dir:
+            training_args = GRPOConfig(
+                output_dir=tmp_dir,
+                per_device_train_batch_size=3,
+                num_generations=3,
+                use_liger_loss=True,
+                max_completion_length=self.max_length,
+                max_steps=4,
```
Comment (qgallouedec) [inline tests/slow/test_grpo_slow.py:53]:
```suggestion
```

**Review by qgallouedec — APPROVED** [review]
Nice!!
